Heap refinement Part3 -- make full use of HP SRAM banks

keyonjie opened the pull request:
This series implements the calculation of the runtime heap at run time, to make full use of HP SRAM banks on Intel cAVS platforms.
  
    Implementation of calculating the runtime heap size at run time, which
    leverage the value of the _runtime_heap_size from platform linker script
    at linking stage.
    
    This will help to make full use of the HP SRAM banks for each cAVS
    platforms, all of them available in .bss section will be used for
    runtime heap allocation.

keyonjie commented:
> `make -C installer/`.

Thanks for information, let me fix it today.

lyakh commented on src/platform/intel/cavs/lib/memory.c:
```diff
@@ -23,6 +23,8 @@ extern uintptr_t _buffer_heap, _sof_core_s_start;
 #if CONFIG_CORE_COUNT > 1
 extern uintptr_t _runtime_shared_heap, _system_shared_heap;
 #endif
+extern uint32_t _runtime_heap_size;
```
`size_t`?

keyonjie replied:
let me change to calculate this size locally.

lyakh commented on src/platform/intel/cavs/lib/memory.c:
```diff
@@ -74,40 +76,40 @@ static SHARED_DATA struct block_map sys_rt_heap_map[CONFIG_CORE_COUNT][3] = {
+static SHARED_DATA struct block_hdr mod_block512[HEAP_COUNT512 * RT_TIMES_MAX];
+static SHARED_DATA struct block_hdr mod_block1024[HEAP_COUNT1024 * RT_TIMES_MAX];
+static SHARED_DATA struct block_hdr mod_block2048[HEAP_COUNT2048 * RT_TIMES_MAX];
+static SHARED_DATA struct block_hdr mod_block4096[HEAP_COUNT4096 * RT_TIMES_MAX];
```
"times" seems a bit unclear - count?

keyonjie replied:
um, we already have HEAP_COUNTxxxx, so.., "times" means multiplying here.

lyakh replied:
yeah, I see it was already used before this patch...

lyakh commented on src/platform/intel/cavs/lib/memory.c:
```diff
@@ -130,6 +132,8 @@ static SHARED_DATA struct mm memmap;
 
 void platform_init_memmap(struct sof *sof)
 {
+	uint32_t heap_runtime_size;
```
`size_t`

lyakh commented on src/platform/intel/cavs/lib/memory.c:
```diff
@@ -130,6 +132,8 @@ static SHARED_DATA struct mm memmap;
 void platform_init_memmap(struct sof *sof)
 {
+	uint32_t heap_runtime_size;
+	uint32_t rt_times;
```
`unsigned int count;` or `n` or something simple

keyonjie commented:
@marc-hb do you have insight about why qemu boot fail on APL here?

marc-hb commented:
> @marc-hb do you have insight about why qemu boot fail on APL here?

No, looks like the firmware just failed to boot. It could be for many reasons. It happened to me some time ago and it was because I was reading un-initialized memory. This  was working (sometimes?) on hardware but not on QEMU.

This is unfortunately difficult to debug because gdb does not work, on the other hand it's very easy to reproduce, just run the same scripts in `.github/workflows` that github runs.

keyonjie commented:
> > @marc-hb do you have insight about why qemu boot fail on APL here?
> 
> No, looks like the firmware just failed to boot. It could be for many reasons. It happened to me some time ago and it was because I was reading un-initialized memory. This was working (sometimes?) on hardware but not on QEMU.

When I just went through the qemu/hw/adsp/dsp/cavs.c, the ADSP_CAVS_1_x_DSP_HP_SRAM_SIZE looks not quite correct there, not sure if it will lead this this boot fail issue.

> 
> This is unfortunately difficult to debug because gdb does not work, on the other hand it's very easy to reproduce, just run the same scripts in `.github/workflows` that github runs. Then it's very quick to try many variations.

Haven't touched the QEMU part for years, @marc-hb do you have a guide about how to build and run it on a non-docker environment?

marc-hb commented:
> When I just went through the qemu/hw/adsp/dsp/cavs.c, the ADSP_CAVS_1_x_DSP_HP_SRAM_SIZE looks not quite correct there, not sure if it will lead this this boot fail issue.

The bad thing about our qemu status is that gdb (where a lot of the qemu value is) does not work right now. The good qemu thing that we still have is that we can still change one line or even one character and retest instantly. Very efficient.

> Haven't touched the QEMU part for years, @marc-hb do you have a guide about how to build and run it on a non-docker environment?

It had been a long time for me too but I just ran the same scripts than CI and it fortunately worked inside and outside docker the same. No guide was required.

Pay attention to the current directory though; the only "comfort" change I made was https://github.com/thesofproject/qemu/pull/46 Appreciate if you can test that change (it's been a while) 

I also tried the sof-5.2 branch in the hope gdb would work but I would not recommend it yet: gdb still does not work, it's more difficult to compile and it's not tested by CI

keyonjie commented:
@marc-hb I realize the APL qemu boot fail happens with gcc built FW only, I have build failure with gcc for APL, looks updating cross-ng doesn't fix the building issue for me...
```
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S: Assembler messages:
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'wur.ae_cw_sd_no'
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'wur.ae_cend0'
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'ae_l64.i'
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'ae_l64.i'
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'ae_l64.i'
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'ae_l64.i'
/home/keyon/github/keyonjie/sof/src/arch/xtensa/xtos/core-restore.S:241: Error: unknown opcode or format name 'ae_l64.i'

```

marc-hb commented:
I don't see any connection between this PR breaking the boot on QEMU and your build failure, is there one?

I have been building APL with the crosstool-ng toolchain a million times over the last year and I never had any issue, neither with gcc 8 nor with the more recent gcc 10. Same for github actions and Jenkins. 

https://thesofproject.github.io/latest/getting_started/build-guide/build-from-scratch.html#build-sof has recently been tested and updated by both @ujfalusi and myself, so you can start that from scratch. Or use docker?

keyonjie commented:
Thanks @marc-hb I managed to run this locally without docker, but looks it boots for me without errors:
```
$ ./xtensa-softmmu/qemu-system-xtensa -cpu broxton -M adsp_bxt -nographic -kernel /home/keyon/hdd/github/keyonjie/sof/build_apl_gcc/src/arch/xtensa/sof-apl.ri -rom /home/keyon/hdd/github/keyonjie/sof/build_apl_gcc/src/arch/xtensa/rom-apl.bin
bridge-io: qemu-bridge-l2-sram-mem fd 9 region 1 at 0x7fb1189d5000 allocated 2097152 bytes
bridge-io: qemu-bridge-hp-sram-mem fd 10 region 2 at 0x7fb118712000 allocated 2097152 bytes
bridge-io: qemu-bridge-lp-sram-mem fd 11 region 3 at 0x7fb120234000 allocated 131072 bytes
bridge-io: qemu-bridge-imr-mem fd 12 region 4 at 0x7fb118692000 allocated 524288 bytes
bridge-io: qemu-bridge-rom-mem fd 13 region 5 at 0x7fb120232000 allocated 8192 bytes
bridge-io: qemu-bridge-cmd-io fd 14 region 6 at 0x7fb120280000 allocated 16 bytes
bridge-io: qemu-bridge-res-io fd 15 region 7 at 0x7fb120231000 allocated 16 bytes
bridge-io: qemu-bridge-ipc-dsp-io fd 16 region 8 at 0x7fb120230000 allocated 32 bytes
bridge-io: qemu-bridge-idc0-io fd 17 region 9 at 0x7fb12022f000 allocated 128 bytes
bridge-io: qemu-bridge-idc1-io fd 18 region 10 at 0x7fb12022e000 allocated 128 bytes
bridge-io: qemu-bridge-hostwin0-io fd 19 region 11 at 0x7fb12022d000 allocated 8 bytes
bridge-io: qemu-bridge-hostwin1-io fd 20 region 12 at 0x7fb118691000 allocated 8 bytes
bridge-io: qemu-bridge-hostwin2-io fd 21 region 13 at 0x7fb118690000 allocated 8 bytes
bridge-io: qemu-bridge-hostwin3-io fd 22 region 14 at 0x7fb11868f000 allocated 8 bytes
bridge-io: qemu-bridge-irq-io fd 23 region 15 at 0x7fb11868e000 allocated 512 bytes
bridge-io: qemu-bridge-timer-io fd 24 region 16 at 0x7fb11868d000 allocated 512 bytes
bridge-io: qemu-bridge-mn-io fd 25 region 17 at 0x7fb11868c000 allocated 512 bytes
bridge-io: qemu-bridge-l2-io fd 26 region 18 at 0x7fb11868b000 allocated 64 bytes
bridge-io: qemu-bridge-ssp0-io fd 27 region 19 at 0x7fb118689000 allocated 8192 bytes
bridge-io: qemu-bridge-ssp1-io fd 28 region 20 at 0x7fb118687000 allocated 8192 bytes
bridge-io: qemu-bridge-ssp2-io fd 29 region 21 at 0x7fb118685000 allocated 8192 bytes
bridge-io: qemu-bridge-ssp3-io fd 30 region 22 at 0x7fb118683000 allocated 8192 bytes
bridge-io: qemu-bridge-ssp4-io fd 31 region 23 at 0x7fb118681000 allocated 8192 bytes
bridge-io: qemu-bridge-ssp5-io fd 32 region 24 at 0x7fb11867f000 allocated 8192 bytes
bridge-io: qemu-bridge-dmac0-lp-io fd 33 region 25 at 0x7fb11867e000 allocated 128 bytes
bridge-io: qemu-bridge-dmac1-hp-io fd 34 region 26 at 0x7fb11867d000 allocated 2048 bytes
bridge-io: qemu-bridge-shim-io fd 35 region 27 at 0x7fb11867c000 allocated 256 bytes
bridge-io: qemu-bridge-gtw-lout-io fd 36 region 28 at 0x7fb11867b000 allocated 448 bytes
bridge-io: qemu-bridge-gtw-lin-io fd 37 region 29 at 0x7fb11867a000 allocated 448 bytes
bridge-io: qemu-bridge-gtw-hout-io fd 38 region 30 at 0x7fb118679000 allocated 896 bytes
bridge-io: qemu-bridge-gtw-hin-io fd 39 region 31 at 0x7fb118678000 allocated 896 bytes
bridge-io-mq: added /qemu-io-parent-bxt
bridge-io-mq: added /qemu-io-child-bxt
now loading:
 kernel /home/keyon/hdd/github/keyonjie/sof/build_apl_gcc/src/arch/xtensa/sof-apl.ri
 ROM /home/keyon/hdd/github/keyonjie/sof/build_apl_gcc/src/arch/xtensa/rom-apl.bin
bridge-io: 0 messages are currently on child queue.
Header $AM1 found at offset 0x2300 bytes
ROM loader: copy 54 kernel pages to IMR
QEMU 4.2.0 monitor - type 'help' for more information
(qemu) 

```

marc-hb commented:
Looks like you're not running the test.

lgirdwood reviewed (changes requested):
@keyonjie I think we do need to change ```_TIMES_``` to ```_COUNT_```, the meaning is slightly confusing for native English speakers (i.e. it may confuse a lot of people about usage).  This should be done in all places.

keyonjie commented:
> @keyonjie I think we do need to change `_TIMES_` to `_COUNT_`, the meaning is slightly confusing for native English speakers (i.e. it may confuse a lot of people about usage). This should be done in all places.

Okay, let me rename that after fix the APL qemu boot failure.

lgirdwood commented:
@kkarask can you check the CI, I cant read the DMA trace logs for the failing test - they are loading forever. Do we need to rerun this test ?
@keyonjie I assume you cant read the DMA log either on the failed test ?

marc-hb commented:
> Okay, let me rename that after fix the APL qemu boot failure.

What was the issue and what is the fix?

keyonjie commented:
> @kkarask can you check the CI, I cant read the DMA trace logs for the failing test - they are loading forever. Do we need to rerun this test ?
> @keyonjie I assume you cant read the DMA log either on the failed test ?

no, I can't neither, let me re-trigger it.

keyonjie commented:
SOFCI TEST

keyonjie commented:
> > Okay, let me rename that after fix the APL qemu boot failure.
> 
> What was the issue and what is the fix?

The debugging shows me that the last 4KB of the HP SRAM is not usable on APL QEMU, it is too difficult to debug the QEMU code now, so I leave a TODO here:

https://github.com/thesofproject/sof/pull/4521/files#diff-e7c6bfd4e93a976c75c0d2cfb9e6dca4c99a3d56ed9cf7188dd3c7fd6216956aR252

keyonjie commented:
> SOFCI TEST

Looks this doesn't trigger  retest of "Internal Intel CI System/merge/build", @kkarask can you help on this?

marc-hb commented:
> > SOFCI TEST

> Looks this doesn't trigger retest of "Internal Intel CI System/merge/build", 

Indeed this is just Jenkins.

> @kkarask can you help on this?

No need as any force push (even an empty `git commit --amend`) will re-trigger everything as long as the git SHA1 has changed.

keyonjie commented:
> > > SOFCI TEST
> 
> > Looks this doesn't trigger retest of "Internal Intel CI System/merge/build",
> 
> Indeed this is just Jenkins.
> 
> > @kkarask can you help on this?
> 
> No need as any force push (even an empty `git commit --amend`) will re-trigger everything as long as the git SHA1 has changed.

Thanks, it works. @lgirdwood the Intelal Intel CI pass this time.

lgirdwood commented:
> The debugging shows me that the last 4KB of the HP SRAM is not usable on APL QEMU, it is too difficult to debug the QEMU code now, so I leave a TODO here:
> 
> https://github.com/thesofproject/sof/pull/4521/files#diff-e7c6bfd4e93a976c75c0d2cfb9e6dca4c99a3d56ed9cf7188dd3c7fd6216956aR252

The stack is in the last 4kB and grows down from top of SRAM.

keyonjie commented:
> > The debugging shows me that the last 4KB of the HP SRAM is not usable on APL QEMU, it is too difficult to debug the QEMU code now, so I leave a TODO here:
> > https://github.com/thesofproject/sof/pull/4521/files#diff-e7c6bfd4e93a976c75c0d2cfb9e6dca4c99a3d56ed9cf7188dd3c7fd6216956aR252
> 
> The stack is in the last 4kB and grows down from top of SRAM.

This is changed in the PR here, see below:

I suspected the same item, and tried move the heap runtime before the stack, but the issue still happened.

BTW, @lgirdwood this is all up to the FW itself, there is no any hard-coding of that in the QEMU side?

```
 * +------------------+-------------------------+-----------------------------+
 * | SOF_FW_START     | text                    |                             |
 * |                  | data                    |                             |
 * |                  | ----------------------- |                             |
 * |                  |+-----------------------+|                             |
 * |                  ||BSS:                   ||                             |
 * |                  ||-----------------------++-----------------------------+
 --* |                  ||Runtime Heap           ||  HEAP_RUNTIME_SIZE          |
 --* |                  ||-----------------------++-----------------------------+
 * |                  ||Runtime shared Heap    ||  HEAP_RUNTIME_SHARED_SIZE   |
 * |                  ||-----------------------++-----------------------------+
 * |                  ||System shared Heap     ||  HEAP_SYSTEM_SHARED_SIZE    |
 * |                  ||-----------------------++-----------------------------+
 * |                  ||Module Buffers         ||  HEAP_BUFFER_SIZE           |
 * |                  ||-----------------------++-----------------------------+
 * |                  ||Primary core Sys Heap  ||  HEAP_SYSTEM_M_SIZE         |
 * |                  ||-----------------------++-----------------------------+
 * |                  ||Pri. Sys Runtime Heap  ||  HEAP_SYS_RUNTIME_M_SIZE    |
 * |                  ||-----------------------++-----------------------------+
 * |                  ||Primary core Stack     ||  SOF_STACK_SIZE             |
 * |                  ||-----------------------++-----------------------------+
 * |                  ||Sec. core Sys Heap     ||  SOF_CORE_S_T_SIZE          |
 * |                  ||Sec. Sys Runtime Heap  ||                             |
 * |                  ||Secondary core Stack   ||                             |
 * |                  | ----------------------- |                             |
 * |                  ||-----------------------++-----------------------------+
 ++* |                  ||Runtime Heap           ||  HEAP_RUNTIME_SIZE          |
 ++* |                  |+-----------------------+|                             |
 * +------------------+-------------------------+-----------------------------+
 */
```

lgirdwood commented:
@keyonjie ah, I see stack has been moved on APL, nothing in qemu that prevents using top 4kB, but please make sure we are not accessing the address at HP SRAM base + HP SRAM size as this is outside the address space.

lgirdwood commented:
SOFCI TEST

keyonjie commented:
> @keyonjie ah, I see stack has been moved on APL, nothing in qemu that prevents using top 4kB, but please make sure we are not accessing the address at HP SRAM base + HP SRAM size as this is outside the address space.

yes, that is the end address so an address access should be smaller than that, this is kept intact with my PR.

mwasko commented:
I am ok with the changes. Please only update naming as @lgirdwood  suggested and get confirmation from @kkarask on FW validation full scope test results.

lgirdwood commented:
@kkarask please let us know when you have the FW validation full scope test result sand if good, we can merge.

lgirdwood commented:
@kkarask ping - any update ?

kkarask commented:
@lgirdwood Sorry for late response, I was on vacation last weeks. I will be back with results.

kkarask commented:
I ran full scope tests and except for known issue one new failure appeared.

Test 12_00_TestMixerPlayback16000Hz16b16b8ch failed with logs:
***
[    43945592.968750] (        5183.072754) c0 pipe         2.8   ....../pipeline-stream.c:252  pipe trigger cmd 0
[    43945606.666667] (          13.697917) c0 ll-schedule        ./schedule/ll_schedule.c:516  task cancel 0xbe160140 pipe-task <f11818eb-e92e-4082-82a3-dc54c604ebb3>
[    43946905.781250] (        1299.114624) c0 pipe         3.11  ....../pipeline-stream.c:252  pipe trigger cmd 2
[    43946917.708333] (          11.927083) c0 ll-schedule        ./schedule/ll_schedule.c:516  task cancel 0xbe1601c0 pipe-task <f11818eb-e92e-4082-82a3-dc54c604ebb3>
[    43946927.473958] (           9.765625) c0 ll-schedule        ./schedule/ll_schedule.c:336  num_tasks 4 total_num_tasks 4
[    43949034.218750] (        2106.744873) c0 pipe         3.11  ....../pipeline-stream.c:252  pipe trigger cmd 0
[    43949049.192708] (          14.973958) c0 ssp-dai      1.0   /drivers/intel/ssp/ssp.c:1059 ssp_trigger() cmd 0
[    43950098.645833] (        1049.453125) c0 wait                         src/lib/wait.c:46   **ERROR poll timeout reg 487496 mask 63 val 0 us 937**
[    43950106.744792] (           8.098958) c0 ssp-dai      1.0   /drivers/intel/ssp/ssp.c:59   WARN ssp_empty_tx_fifo() warning: timeout
[    43950115.833333] (           9.088542) c0 ssp-dai      1.0   /drivers/intel/ssp/ssp.c:1021 ssp_stop(), TX stop
[    43950126.119792] (          10.286458) c0 dw-dma                 src/drivers/dw/dma.c:407  dw_dma_stop(): dma 0 channel 0 stop
[    43950138.359375] (          12.239583) c0 ll-schedule        ./schedule/ll_schedule.c:516  task cancel 0xbe1601c0 pipe-task <f11818eb-e92e-4082-82a3-dc54c604ebb3>
[    43950147.161458] (           8.802083) c0 ll-schedule        ./schedule/ll_schedule.c:516  task cancel 0xbe1600c0 pipe-task <f11818eb-e92e-4082-82a3-dc54c604ebb3>
[    43950157.500000] (          10.338542) c0 ll-schedule        ./schedule/ll_schedule.c:336  num_tasks 3 total_num_tasks 3
[    43952042.734375] (        1885.234375) c0 pipe         4.15  ....../pipeline-stream.c:252  pipe trigger cmd 2
***
It occurred on ICL RVP and WHL RVP Platforms.

I'm not sure if it is related with this change, so I ran additional 300 iterations and issue doesn't appear.

lgirdwood commented:
Thank you @kkarask - good effort. This is a known SSP message and is unrelated.
@keyonjie can you rebase,  we are good to merge,

plbossart commented on src/include/sof/drivers/dw-dma.h:
```diff
@@ -155,10 +150,10 @@ struct dw_lli {
+	 * in case of more than two items
 	 */
-	uint8_t reserved[DW_DMA_LLI_ALIGN - sizeof(uint32_t) * 7];
+	uint32_t reserved;
```
what does this have to do with this PR? this seems completely unrelated to the use of HP SRAM banks?

plbossart commented on src/drivers/dw/dma.c:
```diff
@@ -557,9 +554,8 @@ static int dw_dma_set_config(struct dma_chan_data *channel,
-					sizeof(struct dw_lli) * channel->desc_count,
-					PLATFORM_DCACHE_ALIGN);
+		dw_chan->lli = rballoc(0, SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_DMA,
+				       sizeof(struct dw_lli) * channel->desc_count);
```
again separate PR? This is unrelated to the allocation size, you're just changing how the LLI is handled.

lgirdwood replied:
Looks like the rebase was on the wrong branch. These are already in @keyonjie other PRs

keyonjie replied:
> again separate PR? This is unrelated to the allocation size, you're just changing how the LLI is handled.

Ah, yes, let me put them to PR #4664

plbossart replied:
alignment?

keyonjie replied:
> alignment?

the rballoc() internal will use default PLATFORM_DCACHE_ALIGN alignment, so no need to specify it explicitly.

plbossart reviewed (commented):
In addition to the comments below, can you clarify @keyonjie 
a) if this series allows for bisection? you've broken the patches by platform, but overall is the functionality preserved at each patch?
b) I am not clear on how the memory is handled on TGL-H, which has a different SRAM size from TGL.
Thanks!

keyonjie commented:
> In addition to the comments below, can you clarify @keyonjie
> a) if this series allows for bisection? you've broken the patches by platform, but overall is the functionality preserved at each patch?

This is actually a good point. Since the change is quite fundamental for the heap management, it's not easy to preserve functionality platform by platform, looks I have to squash them to a big one.

> b) I am not clear on how the memory is handled on TGL-H, which has a different SRAM size from TGL.

Oh, we need to handle it similar as ICL/JSL one, BTW, the PR now is not ready yet, I have to test it on all available platforms, marking it draft now...

> Thanks!

keyonjie commented:
> > > Can we remove the allocation patch, and make the necessary changes only for LLI.
> > 
> > 
> > the LLI fix is in another PR, this PR focus on heap refinement only. It doesn't change everything to uncached, and the runtime_shared is keep unchanged also.
> 
> ok, lets do a call on Monday - we can't change lots if items to always be uncache. See [d8a848b#diff-e68a210410f48b326a276f82d551ccf0bd5532feaa9e064b4cc86040c17415f4](https://github.com/thesofproject/sof/commit/d8a848b6ab816a878026c10cfdf3fce94d0575c1#diff-e68a210410f48b326a276f82d551ccf0bd5532feaa9e064b4cc86040c17415f4)
> 
> The intention is to have all shared (cache coherent) objects use teh coherent API.
> 
> 1. Object is created in uncache.
> 2. Object is acquired via lock, invalidated and it's cached alias is returned (minimal hit of cycles to access the spinlock)
> 3. All object access is now cached (no perf impact).
> 4. Access to object is completed and object is released (writeback then unlock).

This sounds a good design to me.

Some opens:
1. buffer shared b/w DMAC and DSP core. we need buffer size >= 2 periods and each period of buffer own by a separated object.
    Since we are using timer not DMA interrupt for period copying, the dmac has to hold the shared object for whole period (e.g. 1ms) that the DSP don't have chance to touch it at all, so we need to make sure buffer is split into periods, and each period of buffer should act as a single object and has its separated lock. Then, e.g. when DMAC is holding the period 0 buffer object and writing to it, the DSP core can hold the period 1 buffer object and reading from it.
2. performance data Vs using uncached memory directly.
    There could be possibility that if the buffer is small enough (e.g. 64Bytes), using this coherent API doesn't increase performance for us at all while consuming extra memory (the coherent header) and invalidation/writeback. If so, we might need a guide of that when should we use the coherent API buffer and should the uncached one should be used.

Yes, let's do a call to talk about it.

lyakh commented:
> > > @lgirdwood @lyakh @kv2019i I am changing the HEAP_RUNTIME_SHARED_SIZE to be finalized at link stage, how should I handle it for Zephyr?
> > 
> > 
> > @keyonjie as discussed in a chat, I'd propose to make the heap buffer in wrapper.c a single array again and split it into two parts at run-time after you've calculated your sizes. Then just use a pointer to the second part of the buffer as the second array is currently used, and adjust sizes respectively.
> 
> @lyakh can you paste what you suggest here? at the moment, the commit "zephyr: wrapper: hardcode HEAP_RUNTIME_SHARED_SIZE to unblock building" is a place holder for that.

that's exactly what I'm proposing: use a single array in .bss and split it into cacheable and non-cacheable at run-time, e.g.
```
static uint8_t __aligned(PLATFORM_DCACHE_ALIGN) heapmem[HEAPMEM_SIZE];

static int statics_init(const struct device *unused)
{
	ARG_UNUSED(unused);

	/* pre_calculated_size_of_uncached_heap == HEAPMEM_SIZE if all is uncached, must be cache-line size aligned */
	size_t size_uncached = pre_calculated_size_of_uncached_heap;

	sys_heap_init(&sof_heap.heap, heapmem, size_uncached);
#ifdef ENABLE_CACHED_HEAP
	sys_heap_init(&sof_heap_shared.heap, heapmem + size_uncached, HEAPMEM_SIZE - size_uncached);
#endif
	return 0;
}

keyonjie commented:
> that's exactly what I'm proposing: use a single array in .bss and split it into cacheable and non-cacheable at run-time, e.g.
> 
> ```
> static uint8_t __aligned(PLATFORM_DCACHE_ALIGN) heapmem[HEAPMEM_SIZE];

The problem is that the total size HEAPMEM_SIZE is not pre-defined, it is calculated by linker script.

keyonjie commented:
@lgirdwood I will update the PR on Wednesday.

keyonjie commented:
@lyakh is it possible to calculate the buffer zone size at linking stage as what I do for the XTOS here?

lyakh commented:
> @lyakh is it possible to calculate the buffer zone size at linking stage as what I do for the XTOS here?

@keyonjie with the present Zephyr approach - no, I don't see a (simple and clean) way to do that precisely. Currently a simple array is used for the heap backing storage on SOF+Zephyr. Its size should be known at linking time. So the only way I can see to really make use of all the available RAM under Zephyr would be to also replace that array with a linker calculated memory area, but that would be a rather large change - we'd need to change 4 linker scripts for cAVS 1.5, 1.8, 2.0 and 2.5. Whereas just fixing your PR to preserve the current state on Zephyr shouldn't be too difficult, maybe we can do just that for now?

lgirdwood commented on src/audio/buffer.c:
```diff
@@ -30,13 +30,6 @@ struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t align)
 	tr_dbg(&buffer_tr, "buffer_alloc()");
 
-	/* validate request */
-	if (size == 0 || size > HEAP_BUFFER_SIZE) {
```
Doesn't size == 0 mean the topology is bad ?

keyonjie replied:
maybe let me keep the 'size==0' check.

lgirdwood commented on src/arch/xtensa/lib/cpu.c:
```diff
@@ -46,7 +47,7 @@ static void alloc_shared_secondary_cores_objects(void)
 	uint8_t *dynamic_vectors;
 
-	dynamic_vectors = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, 0, SOF_DYNAMIC_VECTORS_SIZE);
+	dynamic_vectors = rballoc(SOF_MEM_FLAG_UNCACHE, 0, SOF_DYNAMIC_VECTORS_SIZE);
```
It's better in English to say ```SOF_MEM_COHERENT``` here as it makes the usage more obvious.

keyonjie replied:
> It's better in English to say `SOF_MEM_COHERENT` here as it makes the usage more obvious.

makes sense, let me change.

lgirdwood commented on zephyr/wrapper.c:
```diff
@@ -72,6 +72,9 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
 
+#define HEAP_SYSTEM_SHARED_SIZE	0x1500
+#define HEAP_RUNTIME_SHARED_SIZE	0x10000
+#define HEAP_BUFFER_SIZE	0x10000
```
@lyakh why is this needed by Zephyr at all given that it has it own linker script and allocator ?

keyonjie replied:
yes, @lyakh I need your help here, as it still fails on ICL zephyr building with this hard-coding and it breaks running on device test.

lyakh replied:
@lgirdwood with Zephyr heap is allocated in wrapper.c as an empty static array, i.e. a .bss object. That means it needs a fixed size, known at compile-time. And 64K seems way too little to me for all the allocations on all platforms.

keyonjie replied:
> @lgirdwood with Zephyr heap is allocated in wrapper.c as an empty static array, i.e. a .bss object. That means it needs a fixed size, known at compile-time. And 64K seems way too little to me for all the allocations on all platforms.

@lyakh do you propose to use different values for different platforms via "#ifdef" here?

lyakh replied:
@keyonjie first it would be good to know what heap sizes we're currently using with Zephyr on all platforms. Maybe that information is available somewhere during linking / rimage processing or it should be easy to print it for debugging. Then an obvious safe solution would be to use the same values. But in the long term I think we should also switch over to link-time calculated heap size for Zephyr. For that we could get the linker provide a heap start and heap size address to be used in `statics_init()` in wrapper.c. I'm not a linker expert but I think it should be possible.

aiChaoSONG commented:
SOFCI TEST

RanderWang reviewed (commented):
rzalloc is replced by rballoc.  rzalloc will allocate memory with zero filled, how about rballoc ?

keyonjie commented:
> rzalloc is replced by rballoc. rzalloc will allocate memory with zero filled, how about rballoc ?

yes, it is zeroed in rballoc also, see it here:
```
/* allocates continuous buffers - not for direct use, clients use rballoc() */
void *rballoc_align(uint32_t flags, uint32_t caps, size_t bytes,
		    uint32_t alignment)
{
	struct mm *memmap = memmap_get();
	void *ptr = NULL;
	uint32_t lock_flags;

	spin_lock_irq(&memmap->lock, lock_flags);

	ptr = _balloc_unlocked(flags, caps, bytes, alignment);

	spin_unlock_irq(&memmap->lock, lock_flags);

	if (ptr)
		bzero(ptr, bytes);

	DEBUG_TRACE_PTR(ptr, bytes, SOF_MEM_ZONE_BUFFER, caps, flags);
	return ptr;
}
```

lgirdwood commented on src/lib/alloc.c:
```diff
@@ -875,6 +875,8 @@ static void *alloc_heap_buffer(struct mm_heap *heap, uint32_t flags,
 	return ptr;
 }
 
+#define USE_UNCACHE_BUFFER_ONLY
```
This needs to be a Kconfig option i.e. FORCE allocator to use coherent memory only - and explain it's a debug option.

keyonjie replied:
> This needs to be a Kconfig option i.e. FORCE allocator to use coherent memory only - and explain it's a debug option.

Okay, will do. Do we want to make this "use coherent memory only" selected by default at the moment? As without it, we are hitting kinds of Xrun issues which may annoy the CI a lot.

lgirdwood reviewed (commented):
Most of this PR is good to go in now, except the Zephyr memory size patch and the uncached buffer patch. It may be worthwhile to split the PR into smaller chunks wich can be merged whilst we work through the opens.

keyonjie commented:
> Most of this PR is good to go in now, except the Zephyr memory size patch and the uncached buffer patch. It may be worthwhile to split the PR into smaller chunks wich can be merged whilst we work through the opens.

Just did an debugging and read out the Zephyr memory size for different platforms, pushed the updates.

keyonjie commented:
@lgirdwood the Kconfig item added also.

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
 
 #else
```
How about for IMX (the above #if-else statement?

keyonjie replied:
isn't it intact for the IMX case here?

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
+#elif (CONFIG_HP_MEMORY_BANKS >= 46)
+/* CNL/ICL/TGL */
+#define	HEAP_SIZE	0x200000
+#endif
```
Hmm, this is a shortterm workaround (e.g.  Zephyr's libc already has similar code to figure out the heap size:
zephyr/lib/libc/newlib/libc-hooks.c ), so I'd prefer a very simple definition and add a TODO comment to clearly mark this is a stopgap.

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
-#define HEAPMEM_SHARED_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + \
-				HEAP_RUNTIME_SHARED_SIZE + HEAP_SYSTEM_SHARED_SIZE)
+#if (CONFIG_HP_MEMORY_BANKS <= 8)
+/* APL */
```
I'd remove these comments and add one "/* TODO: heap size should be set based on linker output */" instead at top.

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
+#if (CONFIG_HP_MEMORY_BANKS <= 8)
+/* APL */
+#define	HEAP_SIZE	0x30000
+#elif (CONFIG_HP_MEMORY_BANKS == 16)
```
Add "CONFIG_HP_MEMORY_BANKS >= 16"

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
 
-#define HEAPMEM_SHARED_SIZE	(HEAP_SYSTEM_SIZE + HEAP_RUNTIME_SIZE + \
-				HEAP_RUNTIME_SHARED_SIZE + HEAP_SYSTEM_SHARED_SIZE)
+#if (CONFIG_HP_MEMORY_BANKS <= 8)
```
"CONFIG_HP_MEMORY_BANKS < 16"

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
+#elif (CONFIG_HP_MEMORY_BANKS == 16)
+/* JSL */
+#define	HEAP_SIZE	0x80000
+#elif (CONFIG_HP_MEMORY_BANKS == 30)
```
">= 30"

kv2019i commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,32 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
 #ifdef ENABLE_CACHED_HEAP
-#define HEAPMEM_SIZE		HEAP_BUFFER_SIZE
+/* half heap is cacheable */
+#define	HEAP_SYSTEM_CACHED_SIZE	(HEAP_SIZE / 2)
```
Hmm, this is problematic. We can get size of the heap from linker (and hardcode it in shortterm), but the ratio of the heap that should be allocated to coherent-versus-noncoherent usage is SOF specific information and that we'd have to get from SOF.  We should assume "half is cacheable" here, but rather this should come from some common definition.

kv2019i replied:
Ack @keyonjie . So yeah, then I think we should follow suite with Zephyr and use a single heap as well and serve both cached and non-cached allocations from the same heap area. No need to block this PR to wait for the Zephyr work to complete.

kv2019i reviewed (commented):
The Zephyr heap implementation in wrapper.c is very basic, so I think we can proceed with a fairly simple approach there and focus this review on the XTOS aspects.

lgirdwood commented on Kconfig.xtos-dbg:
```diff
@@ -12,3 +12,16 @@ config DEBUG_BLOCK_FREE
+
+config DEBUG_FORCE_COHERENT_BUFFER
+	bool "Force the allocator to allocate coherent buffer only"
+	default y
```
deafult should be ```n``` atm otherwise we will end up with lost of performance bugs being reported.

lgirdwood reviewed (changes requested):
More comments in the Zephyr part needed so everyone knows this this is temp, but should be good to merge tomorrow.
@kv2019i any reason why the Zephyr heaps are partitioned in cache and uncached ? Can we have one type and do the type conversion in a local wrapper (since its all the same PHY memory) ?

kv2019i commented:
@lgirdwood wrote:
> @kv2019i any reason why the Zephyr heaps are partitioned in cache and uncached ? Can we have one type and do the type conversion in a local wrapper (since its all the same PHY memory) ?

This was done only to ensure cache coherence for the heap metadata as the sys_heap.h implementation stores metadata in-between the allocated chunks, and multiple cores may access the metadata at any given time. 

OTOH, our current cached heap doesn't really work yet as we still observe corruption of metadata (albeit only in long stress tests, but still something is wrong). Once we get a standalone cached heap to work, we could take another look at whether same coherence protocol (for the metadata) could be implemented within context of a single heap. I do agree not doing a static split between the two, would be a much better design.

marc-hb commented:
> More comments in the Zephyr part needed so everyone knows this this is temp, but should be good to merge tomorrow.

Zephyr does not even build right now: https://github.com/zephyrproject-rtos/zephyr/issues/38349

keyonjie commented:
@lgirdwood the interesting thing is that if we are hitting kinds of CI failure with keeping uses cached buffers, please see the latest CI results.

lgirdwood commented:
@keyonjie some red on internal CI, can you split the PR so we can bisect the problematic patch. Some of the updates are good to go now.

keyonjie commented:
@lgirdwood planing to split it to 3, the 1st one here:
https://github.com/thesofproject/sof/pull/4730

plbossart commented on zephyr/wrapper.c:
```diff
@@ -72,18 +72,33 @@ __section(".heap_mem") static uint8_t __aligned(64) heapmem[HEAPMEM_SIZE];
+#define	HEAP_SIZE	0x80000
+#elif (CONFIG_HP_MEMORY_BANKS < 45)
+/* e.g. TGL-H */
+#define	HEAP_SIZE	0x100000
```
can you clarify how this is handled in rimage, your changes only deal with 'tgl' and there's no specific memory size for TGL-H.

keyonjie replied:
> can you clarify how this is handled in rimage, your changes only deal with 'tgl' and there's no specific memory size for TGL-H.

The memory size is configured in the Kconfig, see it here:
https://github.com/thesofproject/sof/blob/8fc5247289db0093bbb9de369c1dcb7e0d19a109/src/arch/xtensa/configs/tgph_defconfig#L8

For rimage, we need to specify image_size according to the memory size, see it here:
https://github.com/thesofproject/rimage/blob/9a26e4558094f19f3f7becd89eb9e8a9a9dd82b9/config/tgl-h.toml#L6

plbossart replied:
but you didn't modify the tgl-h.toml file in this PR, did you? see https://github.com/thesofproject/sof/pull/4521/commits/1b0757e01555c01cc288a5182571a919620dde1c

keyonjie replied:
> but you didn't modify the tgl-h.toml file in this PR, did you? see [1b0757e](https://github.com/thesofproject/sof/commit/1b0757e01555c01cc288a5182571a919620dde1c)

That's because the tgl-h.toml is already updated before my PR, the rimage is already updated to 9a26e4558094 in the sof upstream repo:

9a26e4558094 config: apl: change image_size to the real SRAM size
8a2ea0044dcf config: cnl: change image_size to the real SRAM size
1de9ccca3388 config: jsl: change image_size to the real SRAM size
f52a078aadf1 config: icl: change image_size to the real SRAM size
8073ea3ef93c config: tgl-h: change image_size to the real SRAM size
9e50a02f1b0d config: tgl: change image_size to the real SRAM size

plbossart commented on test/cmocka/include/mock_memory.h:
```diff
@@ -6,10 +6,8 @@
+#define HEAP_SYS_RUNTIME_0_BASE 0xBE120000
+#define SOF_CORE_S_START	0xBE140000
+#define HEAP_RUNTIME_BASE	0xBE180000
+#define HEAP_BUFFER_BASE	0xBE1C0000
```
not sure how anyone could review such changes without any comments on how this is inferred?

plbossart commented on src/audio/mux/mux.c:
```diff
@@ -173,7 +173,8 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
 
-	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+	/* allocate quite a big buffer, use rballoc to make sure it will succeed */
+	cd = rballoc(0, SOF_MEM_CAPS_RAM,
```
I don't see why we special-case a component? something's fishy here. why not the mixer as well then?

keyonjie replied:
> I don't see why we special-case a component? something's fishy here. why not the mixer as well then?

The size of 'cd' here is > 1KB, we have very limited runtime blocks on APL, so any ask for allocation size >= 1KB are suggested to use the rballoc() and allocation will happen on the BUFFER zone.

The mixer 'cd' size is very small, using rzalloc() to allocate e.g. a 64Bytes buffer could help us to save memory usage.

lgirdwood replied:
Our allocator is simple and showing its limitations. The Zephyr allocator is far better.

keyonjie commented:
@lgirdwood I am splitting the biggest commit 'alloc: heap memory refinement ' to make life easier, will submit some low hanging fruits first.

keyonjie commented:
let me close this one and go with smaller PRs.
